### PR TITLE
Embed nixpkgs-fmt, refactor to use lsp-server

### PR DESCRIPTION
**Pull request by jD91mZM2**

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1,425 +1,518 @@
...
 
 [[package]]
-name = "rnix"
-version = "0.5.1"
-source = "registry+https://github.com/rust-lang/crates.io-index"
+name = "rnix-lsp"
+version = "0.1.0"
 dependencies = [
+ "env_logger 0.6.2 (registry+https://github.com/rust-lang/crates.io-index)",
  "failure 0.1.5 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (matklad) [inline Cargo.lock]: Note that failure now depends on old `syn`.

File: Cargo.toml [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
-url = "1.7.2"
+log = "0.4.8"
+lsp-server = "0.2.0"
+lsp-types = "0.61.0"
+nixpkgs-fmt = { git = "https://github.com/nix-community/nixpkgs-fmt", rev = "064e542d4b2077b5ba23e9162c8261fb1eb10824" }
+rnix = "0.6.0"
+serde = "1.0.99"
+serde_derive = "1.0.99"
+serde_json = "1.0.40"
+url = "2.1.0"
```
Comment (matklad) [inline Cargo.toml]: I *think* `lsp-types` re-exports `Url`, so there's no need for an explicit dep

File: Cargo.toml [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
-serde_derive = "1.0.85"
-serde_json = "1.0.37"
-url = "1.7.2"
+log = "0.4.8"
+lsp-server = "0.2.0"
+lsp-types = "0.61.0"
+nixpkgs-fmt = { git = "https://github.com/nix-community/nixpkgs-fmt", rev = "064e542d4b2077b5ba23e9162c8261fb1eb10824" }
+rnix = "0.6.0"
+serde = "1.0.99"
+serde_derive = "1.0.99"
```
Comment (matklad) [inline Cargo.toml]: it's better to use `serder` with `derive` feature

File: rust-toolchain [hunk 4]
```diff
@@ -1 +1 @@
-nightly-2019-04-03
+nightly-2019-08-25
```
Comment (matklad) [inline rust-toolchain]: There's no need for nightly I think?

File: src/main.rs [hunk 5]
```diff
@@ -1,183 +1,153 @@
...
-#[macro_use] extern crate failure;
 #[macro_use] extern crate serde_derive;
 
-mod format;
 mod lookup;
 mod models;
 mod utils;
 
-use self::models::*;
+use models::ExtendSelectionParams;
```
Comment (matklad) [inline src/main.rs]: This should be moved to `selectionRanges` API: https://github.com/gluon-lang/lsp-types/blob/bc714cd8836c8bb0f910cd32e87b98c09c721944/src/request.rs#L610-L618

File: src/main.rs [hunk 6]
```diff
@@ -1,183 +1,153 @@
...
 
-    let stdout = io::stdout();
-    let mut app = App {
-        files: HashMap::new(),
-        log: &mut log,
-        stdout: stdout.lock()
-    };
+pub const DUMMY_ERROR: &str = "A fatal error has occured and nix-lsp has shut down.";
+
+fn main() -> Result<(), &'static str> {
```
Comment (matklad) [inline src/main.rs]: Note that `main() -> Result<(), &str>` will use `fmt::Debug` to print `&str`, so, for best results, it's still better to print the error manually.

File: src/main.rs [hunk 7]
```diff
@@ -1,183 +1,153 @@
...
+            trace!("Message: {:#?}", msg);
+            match msg {
+                Message::Request(req) => {
+                    let id = req.id.clone();
+                    match self.conn.handle_shutdown(&req) {
+                        Ok(true) => break,
+                        Ok(false) => if let Err(err) = self.handle_request(req) {
+                            self.err(id, err);
+                        },
+                        Err(err) => self.err(id, err),
```
Comment (matklad) [inline src/main.rs]: It's better to break out of the loop in this case: protocol error during shutdown still means that you should shut down

File: src/main.rs [hunk 8]
```diff
@@ -1,183 +1,153 @@
...
-            "textDocument/didChange" => {
-                let params: DidChangeTextDocumentParams = serde_json::from_value(req.params)?;
-                if let Some(change) = params.content_changes.into_iter().last() {
-                    let parsed = rnix::parse(&change.text);
-                    self.send_diagnostics(params.text_document.uri.clone(), &change.text, &parsed)?;
-                    self.files.insert(params.text_document.uri, (parsed, change.text));
-                }
-            },
-            "textDocument/definition" => {
+            GotoDefinition::METHOD => {
```
Comment (matklad) [inline src/main.rs]: the extraction could be done in a type-safer way: https://github.com/rust-analyzer/rust-analyzer/blob/f90a88630195a1b6b542b5883295d66dd6c59cef/crates/ra_lsp_server/src/main_loop.rs#L521-L540

File: Cargo.lock (hunk 1 above)
Comment (jD91mZM2) [reply Cargo.lock]: Oh, I removed failure so I should've removed it from the Cargo.toml. Oopsie

File: src/main.rs (hunk 6 above)
Comment (jD91mZM2) [reply src/main.rs]: Right, I'm mainly wishing to return the right exit code. I could do it using `process::exit` but... destructors :(

File: src/main.rs (hunk 7 above)
Comment (jD91mZM2) [reply src/main.rs]: Errors only happen inside `handle_shutdown` if it actually detected a shutdown request?

File: src/main.rs (hunk 6 above)
Comment (matklad) [reply src/main.rs]: I like how ripgrep does this, very tidy: https://github.com/BurntSushi/ripgrep/blob/master/src/main.rs#L48

File: src/main.rs (hunk 7 above)
Comment (matklad) [reply src/main.rs]: Yup

Comment (matklad) [reply src/main.rs]: Specifically, if shutdown request was not followed by exit notification

File: src/main.rs (hunk 6 above)
Comment (jD91mZM2) [reply src/main.rs]: That's somewhat what I wanted to avoid, but fine :)

**Comment by jD91mZM2** [comment]
I think the CI failure isn't real, so I'll just merge this and potentially fix stuff later

**Pull request merged**